### PR TITLE
Multiple domains for organization

### PR DESCRIPTION
**Pull request by vbelinschi**
## References

* #0000

### CODE REVIEW

File: source/marketparticipant/Energinet.DataHub.MarketParticipant.ApplyDBMigrationsApp/Scripts/ALL/Model/202410211100 Organization Domain table.sql [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ALTER TABLE [dbo].[OrganizationDomain] ADD
+
+    PeriodStart DATETIME2 GENERATED ALWAYS AS ROW START HIDDEN
+    CONSTRAINT DF_OrganizationDomain_PeriodStart DEFAULT SYSUTCDATETIME(),
+    PeriodEnd DATETIME2 GENERATED ALWAYS AS ROW END HIDDEN
+    CONSTRAINT DF_OrganizationDomain_PeriodEnd   DEFAULT CONVERT(DATETIME2, '9999-12-31 23:59:59.9999999'),
+    PERIOD FOR SYSTEM_TIME(PeriodStart, PeriodEnd),
+    
+    Version             int              NOT NULL DEFAULT 0,
+    ChangedByIdentityId uniqueidentifier NOT NULL
```
Comment (defectiveAi) [inline source/marketparticipant/Energinet.DataHub.MarketParticipant.ApplyDBMigrationsApp/Scripts/ALL/Model/202410211100 Organization Domain table.sql:21]: Need support for DeletedByIdentityId, for when a domain is removed.

File: source/marketparticipant/Energinet.DataHub.MarketParticipant.ApplyDBMigrationsApp/Scripts/ALL/Model/202410211100 Organization Domain table.sql [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    CONSTRAINT DF_ChangedByIdentityId DEFAULT('00000000-FFFF-FFFF-FFFF-000000000000');
+GO
+
+ALTER TABLE [dbo].[OrganizationDomain]
+    SET (SYSTEM_VERSIONING = ON (HISTORY_TABLE = dbo.[OrganizationDomainHistory]))
+GO
+
+INSERT INTO [dbo].[OrganizationDomain](Id, OrganizationId, Domain, Version, ChangedByIdentityId)
+SELECT NEWID(), Id, Domain, Version, ChangedByIdentityId
```
Comment (defectiveAi) [inline source/marketparticipant/Energinet.DataHub.MarketParticipant.ApplyDBMigrationsApp/Scripts/ALL/Model/202410211100 Organization Domain table.sql:30]: Is Version not just 1 always during migration?

File: source/marketparticipant/Energinet.DataHub.MarketParticipant.Domain/Model/OrganizationDomain.cs [hunk 3]
```diff
@@ -35,7 +37,12 @@ public OrganizationDomain(string value)
 
     public string Value { get; }
 
-    public static bool IsValid(string value)
+    public static bool IsValid(IEnumerable<string> values)
```
Comment (defectiveAi) [inline source/marketparticipant/Energinet.DataHub.MarketParticipant.Domain/Model/OrganizationDomain.cs:40]: I would probably keep the old singular API, then use All() whenever multiple domains need validation.

File: source/marketparticipant/Energinet.DataHub.MarketParticipant.Infrastructure/Persistence/Model/OrganizationDomainEntity.cs [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using Energinet.DataHub.MarketParticipant.Infrastructure.Persistence.Audit;
+
+namespace Energinet.DataHub.MarketParticipant.Infrastructure.Persistence.Model;
+
+public sealed class OrganizationDomainEntity : IAuditedEntity
```
Comment (defectiveAi) [inline source/marketparticipant/Energinet.DataHub.MarketParticipant.Infrastructure/Persistence/Model/OrganizationDomainEntity.cs:20]: There is another interface for supporting deletion.

File: source/marketparticipant/Energinet.DataHub.MarketParticipant.Infrastructure/Persistence/Repositories/OrganizationAuditLogRepository.cs [hunk 5]
```diff
@@ -29,13 +30,21 @@ public OrganizationAuditLogRepository(IMarketParticipantDbContext context)
...
-        return new AuditLogBuilder<OrganizationAuditedChange, OrganizationEntity>(dataSource)
+        var organizationDomainChanges = (await new AuditLogBuilder<OrganizationAuditedChange, OrganizationDomainEntity>(organizationDomainDataSource)
+            .Add(OrganizationAuditedChange.Domain, entity => string.Join(";", entity.Domain), AuditedChangeCompareAt.Creation)
+            .BuildAsync().ConfigureAwait(false))
+            .ToList();
+
+        var organizationDataSource = new HistoryTableDataSource<OrganizationEntity>(_context.Organizations, entity => entity.Id == organization.Value);
+
+        var organizationNameChanges = await new AuditLogBuilder<OrganizationAuditedChange, OrganizationEntity>(organizationDataSource)
             .Add(OrganizationAuditedChange.Name, entity => entity.Name, AuditedChangeCompareAt.Creation)
```
Comment (defectiveAi) [inline source/marketparticipant/Energinet.DataHub.MarketParticipant.Infrastructure/Persistence/Repositories/OrganizationAuditLogRepository.cs:45]: How do we get the old audits out, for domain changes on previous table?

File: source/marketparticipant/Energinet.DataHub.MarketParticipant.Infrastructure/Persistence/Repositories/OrganizationAuditLogRepository.cs [hunk 6]
```diff
@@ -29,13 +30,21 @@ public OrganizationAuditLogRepository(IMarketParticipantDbContext context)
...
 
-    public Task<IEnumerable<AuditLog<OrganizationAuditedChange>>> GetAsync(OrganizationId organization)
+    public async Task<IEnumerable<AuditLog<OrganizationAuditedChange>>> GetAsync(OrganizationId organization)
     {
-        var dataSource = new HistoryTableDataSource<OrganizationEntity>(_context.Organizations, entity => entity.Id == organization.Value);
+        var organizationDomainDataSource = new HistoryTableDataSource<OrganizationDomainEntity>(_context.OrganizationDomains, entity => entity.OrganizationId == organization.Value);
 
-        return new AuditLogBuilder<OrganizationAuditedChange, OrganizationEntity>(dataSource)
+        var organizationDomainChanges = (await new AuditLogBuilder<OrganizationAuditedChange, OrganizationDomainEntity>(organizationDomainDataSource)
+            .Add(OrganizationAuditedChange.Domain, entity => string.Join(";", entity.Domain), AuditedChangeCompareAt.Creation)
```
Comment (defectiveAi) [inline source/marketparticipant/Energinet.DataHub.MarketParticipant.Infrastructure/Persistence/Repositories/OrganizationAuditLogRepository.cs:38]: There are examples for adding deletion support.

File: source/marketparticipant/Energinet.DataHub.MarketParticipant.IntegrationTests/Repositories/OrganizationRepositoryTests.cs [hunk 7]
```diff
@@ -237,4 +239,81 @@ await _fixture.PrepareActorAsync(
...
+        var newOrg = await orgRepository2.GetAsync(orgId.Value);
+
+        // Assert
+        Assert.NotNull(newOrg);
+        Assert.NotEqual(Guid.Empty, newOrg.Id.Value);
+        Assert.Equal(testOrg.Domains.Count(), domains.Length);
+    }
+
+    [Fact]
+    public async Task AddOrUpdateAsync_OneOrganizationWitsDuplicateDomains_ReturnsError()
```
Comment (defectiveAi) [inline source/marketparticipant/Energinet.DataHub.MarketParticipant.IntegrationTests/Repositories/OrganizationRepositoryTests.cs:267]: nitpick: Wits > With

File: source/marketparticipant/Energinet.DataHub.MarketParticipant.Domain/Model/Organization.cs [hunk 8]
```diff
@@ -39,22 +40,22 @@ public Organization(
...
+        Domains = domains;
         _organizationStatusTransitioner = new OrganizationStatusTransitioner(status);
     }
 
     public OrganizationId Id { get; }
     public string Name { get; set; }
     public BusinessRegisterIdentifier BusinessRegisterIdentifier { get; }
     public Address Address { get; }
-    public OrganizationDomain Domain { get; set; }
+    public IEnumerable<OrganizationDomain> Domains { get; set; }
```
Comment (defectiveAi) [inline source/marketparticipant/Energinet.DataHub.MarketParticipant.Domain/Model/Organization.cs:58]: Consider if it makes sense to make this HashSet or interface of one.

**Review by defectiveAi — CHANGES_REQUESTED** [review]
Mostly unsure if add/delete of domains is correctly audited + if/how audit of old values is kept. Some tests will show that.

File: source/marketparticipant/Energinet.DataHub.MarketParticipant.ApplyDBMigrationsApp/Scripts/ALL/Model/202410211100 Organization Domain table.sql (hunk 1 above)
Comment (vbelinschi) [reply source/marketparticipant/Energinet.DataHub.MarketParticipant.ApplyDBMigrationsApp/Scripts/ALL/Model/202410211100 Organization Domain table.sql:21]: done

File: source/marketparticipant/Energinet.DataHub.MarketParticipant.ApplyDBMigrationsApp/Scripts/ALL/Model/202410211100 Organization Domain table.sql (hunk 2 above)
Comment (vbelinschi) [reply source/marketparticipant/Energinet.DataHub.MarketParticipant.ApplyDBMigrationsApp/Scripts/ALL/Model/202410211100 Organization Domain table.sql]: ok, I thought we should keep the versions

File: source/marketparticipant/Energinet.DataHub.MarketParticipant.Domain/Model/OrganizationDomain.cs (hunk 3 above)
Comment (vbelinschi) [reply source/marketparticipant/Energinet.DataHub.MarketParticipant.Domain/Model/OrganizationDomain.cs]: done

File: source/marketparticipant/Energinet.DataHub.MarketParticipant.Infrastructure/Persistence/Model/OrganizationDomainEntity.cs (hunk 4 above)
Comment (vbelinschi) [reply source/marketparticipant/Energinet.DataHub.MarketParticipant.Infrastructure/Persistence/Model/OrganizationDomainEntity.cs]: done

File: source/marketparticipant/Energinet.DataHub.MarketParticipant.Infrastructure/Persistence/Repositories/OrganizationAuditLogRepository.cs (hunk 5 above)
Comment (vbelinschi) [reply source/marketparticipant/Energinet.DataHub.MarketParticipant.Infrastructure/Persistence/Repositories/OrganizationAuditLogRepository.cs:49]: I added sql script to move the old logs to the new table as well

File: source/marketparticipant/Energinet.DataHub.MarketParticipant.Infrastructure/Persistence/Repositories/OrganizationAuditLogRepository.cs (hunk 6 above)
Comment (vbelinschi) [reply source/marketparticipant/Energinet.DataHub.MarketParticipant.Infrastructure/Persistence/Repositories/OrganizationAuditLogRepository.cs]: done

File: source/marketparticipant/Energinet.DataHub.MarketParticipant.IntegrationTests/Hosts/WebApi/GetOrganizationAuditLogsHandlerIntegrationTests.cs [hunk 9]
```diff
@@ -82,7 +82,7 @@ public Task GetAuditLogs_ChangeDomain_IsAudited()
             },
```
[the remark this answers is not in the document]
Comment (vbelinschi) [reply source/marketparticipant/Energinet.DataHub.MarketParticipant.IntegrationTests/Hosts/WebApi/GetOrganizationAuditLogsHandlerIntegrationTests.cs:82]: added

File: source/marketparticipant/Energinet.DataHub.MarketParticipant.IntegrationTests/Repositories/OrganizationRepositoryTests.cs (hunk 7 above)
Comment (vbelinschi) [reply source/marketparticipant/Energinet.DataHub.MarketParticipant.IntegrationTests/Repositories/OrganizationRepositoryTests.cs]: 🙈

File: source/marketparticipant/Energinet.DataHub.MarketParticipant.Domain/Model/Organization.cs (hunk 8 above)
Comment (vbelinschi) [reply source/marketparticipant/Energinet.DataHub.MarketParticipant.Domain/Model/Organization.cs:58]: I changed the OrganizationDomain to record, don't know if we get anything from changing to HashSet as well

**Review by defectiveAi — APPROVED** [review]

File: source/marketparticipant/Energinet.DataHub.MarketParticipant.Domain/Model/Organization.cs (hunk 8 above)
Comment (defectiveAi) [reply source/marketparticipant/Energinet.DataHub.MarketParticipant.Domain/Model/Organization.cs:58]: HashSet may give a simpler difference comparison when updaring the entity, assuming EF handles it correctly.

Comment (vbelinschi) [reply source/marketparticipant/Energinet.DataHub.MarketParticipant.Domain/Model/Organization.cs:58]: I started to change it to HashSet, but I think it's too much work for little gain.
Fx.: in `UpdateOrganizationHandler` we can't just set `organization.Domain = newDomains`, we have to clear the HashSet and add the new ones to it.

**Pull request merged**